pkg/lib/controller-runtime/client/ssa.go

```diff
@@ -0,0 +1,202 @@
+package client
+
+import (
+	"context"
+	"fmt"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	kscheme "k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	apiregistrationv1 "k8s.io/kube-aggregator/pkg/apis/apiregistration/v1"
+	"reflect"
+	k8scontrollerclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	defaultOwner = "olm.registry"
+)
+
+type Object interface {
+	runtime.Object
+	metav1.Object
+}
+
+func NewForConfig(c *rest.Config, scheme *runtime.Scheme, owner string) (*ServerSideApplier, error) {
+	if scheme == nil {
+		scheme = runtime.NewScheme()
+		localSchemeBuilder := runtime.NewSchemeBuilder(
+			kscheme.AddToScheme,
+			apiextensionsv1.AddToScheme,
+			apiregistrationv1.AddToScheme,
+		)
+		if err := localSchemeBuilder.AddToScheme(scheme); err != nil {
+			return nil, err
+		}
+	}
+	restClient, err := k8scontrollerclient.New(c, k8scontrollerclient.Options{
+		Scheme: scheme,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if len(owner) == 0 {
+		owner = defaultOwner
+	}
+
+	return &ServerSideApplier{
+		Client: restClient,
+		Scheme: scheme,
+		Owner:  k8scontrollerclient.FieldOwner(owner),
+	}, nil
+}
+
+func SetDefaultGroupVersionKind(obj Object, s *runtime.Scheme) {
+	gvk := obj.GetObjectKind().GroupVersionKind()
+	if gvk.Empty() && s != nil {
+		// Best-effort guess the GVK
+		gvks, _, err := s.ObjectKinds(obj)
+		if err != nil {
+			panic(fmt.Sprintf("unable to get gvks for object %T: %s", obj, err))
+		}
+		if len(gvks) == 0 || gvks[0].Empty() {
+			panic(fmt.Sprintf("unexpected gvks registered for object %T: %v", obj, gvks))
+		}
+		// TODO: The same object can be registered for multiple group versions
+		// (although in practise this doesn't seem to be used).
+		// In such case, the version set may not be correct.
+		gvk = gvks[0]
+	}
+	obj.GetObjectKind().SetGroupVersionKind(gvk)
+}
+
+type ServerSideApplier struct {
+	k8scontrollerclient.Client
```

Comment:
Is there a reason for embedding this? Do we plan on using `ServerSideApplier` as a unified client everywhere?

Reply:
This was mainly so that the testcontext could use its k8s controller client as the `ServerSideApplier`. We can create a separate one just for doing the apply.